File: day2/solve.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>

int part1(void) {
    std::ifstream file{"input.txt"};
    int no_safe = 0;

    for (std::string line; std::getline(file, line); ) {
        bool increasing = true, decreasing = true, safe = true;
        std::istringstream ss(line);
        int prev;
        ss >> prev;

        for (int num; ss >> num;) {
            int diff = num - prev;
            if (increasing && diff >= 1 && diff <= 3) {
                decreasing = false;
            } else if (decreasing && diff <= -1 && diff >= -3) {
                increasing = false;
            } else {
                safe = false;
                break;
            }
            prev = num;
        }
        no_safe += safe;
    }

    return no_safe;
}

int part2(char* input, char* output) {
    int no_safe = 0;

    std::ifstream file{input};    
    for (std::string line; std::getline(file, line); ) {
        std::vector<int> ids;
        std::istringstream ss(line);
        for (int id; ss >> id; ) {
            ids.push_back(id);
        }

        bool safeLine = false;
        auto consider = [&](int j){
            std::vector<int> b = ids;
            b.erase(b.begin() + j);
            bool increasing = true, decreasing = true, safe = true;
            for (int i=0; i<b.size()-1; ++i) {
                int diff = b[i+1] - b[i];
                if (increasing && diff >= 1 && diff <= 3) {
                    decreasing = false;
                } else if (decreasing && diff <= -1 && diff >= -3) {
                    increasing = false;
                } else {
                    safe = false;
                    break;
                }
            }
            if (safe) safeLine |= true;
        };

        consider(0);
        for (int i=0; i<ids.size()-1; ++i) {
            int diff = ids[i+1] - ids[i];
            if (abs(diff) < 1 || abs(diff) > 3){
                consider(i);
                consider(i+1);
                break;
            }
            if (i < ids.size()-2) {
                int diff2 = ids[i+2] - ids[i+1];
                if (diff * diff2 < 0) {
                    consider(i);
                    consider(i+1);
                    consider(i+2);
                    break;
                }
            }
        }
        no_safe += safeLine;
    }

    return no_safe;
}

int main(int argc, char* argv[]) {
    std::cout << "Part 1: " << part1() << "\n";
    std::cout << "Part 2: " << part2(argv[1], argv[2]) << "\n";
    return 0;
}

File: day2/solve.py
def valid_sequence(ids):
    increasing = True; decreasing = True; safe = True
    for i in range(1, len(ids)):
        if increasing and ids[i] >= ids[i-1]+1 and ids[i] <= ids[i-1]+3:
            decreasing = False
        elif decreasing and ids[i] <= ids[i-1]-1 and ids[i] >= ids[i-1]-3:
            increasing = False
        else:
            safe = False
            break
    return safe

def part1():
    file = open("input.txt")
    safe_ids = 0
    for line in file:
        ids = list(map(int, line.split()))
        safe_ids += valid_sequence(ids)
    print("Safe reports:", safe_ids)

def part2():
    file = open("input.txt")
    safe_ids = 0
    for line in file:
        ids = list(map(int, line.split()))
        if not valid_sequence(ids):
            for i in range(0, len(ids)):
                new_ids = [ids[j] for j in range(0, len(ids)) if i != j]
                if valid_sequence(new_ids):
                    safe_ids += 1
                    break
        else:
            safe_ids += 1
    print("Safe reports with 1 replacement: ", safe_ids)

if __name__ == "__main__":
    part1()
    part2()